### PR TITLE
Retries for GetDotNetInstallScript in powershell+bash

### PR DESCRIPTION
**Pull request by vzarytovskii**
Added retries with exponential backoff for any issues Invoke-WebRequest can potentially hit (PowerShell).
Added retries with exponential backoff for any issues curl and wget can potentially hit (bash).
Added response logging, headers, and status code (bash).
Added exception logging when retry happens (Powershell, bash).
Addresses #4740, and potentially workarounds any new instances of [#8687](https://github.com/dotnet/core-eng/issues/8687)

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Unfortunately, it won't be that easy to have diagnostics on par with curl, with the version of cmdlet/Powershell we use.
In the earlier versions, Invoke-WebRequest doesn't expose much and doesn't give you much freedom.

I have added basic errors/exceptions logging on retries. If something more verbose/sophisticated is needed, I would create a separate issue for that, and migrate from Invoke-WebRequest to something more usable.

**Comment by vzarytovskii** [comment]
Right, will take a look at it and include in the same PR

**Comment by vzarytovskii** [comment]
Please, note that I have added more verbosity to both curl and wget, as requested.

File: eng/common/tools.sh [hunk 1]
```diff
@@ -218,6 +218,28 @@ function InstallDotNet {
...
+    fi
+
+    timeout=$((2**$retries-1))
+    echo "Failed to execute '$@'. Waiting $timeout seconds before next attempt ($retries out of $maxRetries)." 1>&2
+    sleep $timeout
+  done
+
+  echo "Failed to execute '$@' for $maxRetries times." 1>&2
+
+  return -1
```
Comment (tkapin) [inline eng/common/tools.sh]: Please return 1

File: eng/common/tools.sh [hunk 2]
```diff
@@ -218,6 +218,28 @@ function InstallDotNet {
   }
 }
 
+function with_retries {
+  local maxRetries=${RETRIES-5}
```
Comment (tkapin) [inline eng/common/tools.sh]: Please throw the env var modificaiton from here.

File: eng/common/tools.ps1 [hunk 3]
```diff
@@ -205,7 +205,7 @@ function GetDotNetInstallScript([string] $dotnetRoot) {
         break
       }
       catch {
-        Write-Message "Failed to download '$uri'"
+        Write-Host "Failed to download '$uri'"
         Write-Error $_.Exception.Message -ErrorAction Continue
```
[the remark this answers is not in the document]
Comment (vzarytovskii) [reply eng/common/tools.ps1]: It will fail if not specified, due to `$ErrorActionPreference = "Stop"`

**Pull request merged**